sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -219,21 +224,80 @@ _az_client_secret_credential_ensure_token_credential(az_client_secret_credential
+    az_span const request_url,
+    az_span_builder * const p_builder) {
+  az_url url = { 0 };
+  if (!az_succeeded(az_url_parse(request_url, &url))) {
+    return AZ_ERROR_ARG;
+  }
+
+  az_span domains[3] = { 0 };
+  size_t const ndomains = AZ_ARRAY_SIZE(domains);
+  for (size_t i = 0; i < ndomains; ++i) {
+    if (!az_succeeded(az_host_read_domain(&url.authority.host, &domains[(ndomains - 1) - i]))) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  // Add "https://"
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, url.scheme));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, AZ_STR("://")));
+
+  for (size_t i = 0; i < (ndomains - 1); ++i) { // This loop would add "vault.azure."
+    AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, domains[i]));
+    AZ_RETURN_IF_FAILED(az_span_builder_append_byte(p_builder, '.'));
+  }
+
+  // We have to do this out of the loop so that we won't append an extra "." at the end.
+  // So this expression is going to add the final "net" to an existing "https://vault.azure."
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, domains[ndomains - 1]));
+
+  return AZ_OK;
+}
+
 static AZ_NODISCARD az_result _az_client_secret_credential_credential_func(
     az_client_secret_credential * const self,
     az_http_request_builder * const hrb) {
   AZ_CONTRACT_ARG_NOT_NULL(self);
   AZ_CONTRACT_ARG_NOT_NULL(hrb);
 
+  az_span const request_url = az_span_builder_result(&hrb->url_builder);
+
   {
     az_url url = { 0 };
-    AZ_RETURN_IF_FAILED(az_url_parse(az_span_builder_result(&hrb->url_builder), &url));
+    AZ_RETURN_IF_FAILED(az_succeeded(az_url_parse(request_url, &url)));
     if (!az_span_eq_ascii_ignore_case(url.scheme, AZ_STR("https"))) {
       return AZ_ERROR_ARG;
     }
   }
 
-  AZ_RETURN_IF_FAILED(_az_client_secret_credential_ensure_token_credential(self));
+  uint8_t request_resource_buf[_az_TOKEN_CREDENTIAL_AUTH_RESOURCE_BUF_SIZE] = { 0 };
+  az_span_builder resource_builder
+      = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(request_resource_buf));
+
+  AZ_RETURN_IF_FAILED(_az_client_secret_credential_get_resource(request_url, &resource_builder));
+  az_span const request_resource = az_span_builder_result(&resource_builder);
+
+  if (az_span_eq(
+          request_resource,
+          (az_span){
+              .begin = self->_token_credential._resource_buf,
+              .size = self->_token_credential._resource_size,
+          })) {
+    AZ_RETURN_IF_FAILED(_az_client_secret_credential_ensure_token_credential(self));
+  } else {
+    resource_builder = az_span_builder_create(
+        (az_mut_span)AZ_SPAN_FROM_ARRAY(self->_token_credential._resource_buf));
+
+    AZ_RETURN_IF_FAILED(az_span_builder_append(&resource_builder, request_resource));
+    self->_token_credential._resource_size = az_span_builder_result(&resource_builder).size;
+
+    AZ_RETURN_IF_FAILED(_az_client_secret_credential_renew_token_credential(self));
+  }
```

Comment:
Without any mutex around this it introduces race conditions if a credential is used by different clients with different resource strings. If different clients are attempting to retrieve tokens for different resources it's possible that one would get a token for the wrong resource. So it would require any callers to call this inside a critical section. However, this becomes a moot point if we refactor as suggested in the comments on https://github.com/Azure/azure-sdk-for-c/pull/250.

Reply:
I'll complete this PR so that it goes to my branch with auth refactoring, and then change that branch according to your suggestions.